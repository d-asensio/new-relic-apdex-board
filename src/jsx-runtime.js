export default function createElement (tag, props, ...children) {
  const element = document.createElement(tag)

  if (props !== null) {
    _addPropsToElement(props, element)
  }

  _appendChildrenToElement(children, element)

  return element
}

function _addPropsToElement (props, element) {
  const { className, ...attributes } = props

  _addAttributesToElement(attributes, element)

  if (className) {
    _addClassNameToElement(className, element)
  }
}

function _addAttributesToElement (attributes, element) {
  const attributeEntries = Object.entries(attributes)

  for (const [attrName, attrValue] of attributeEntries) {
    _attributeSupportedByElementOrThrow(attrName, element)

    element.setAttribute(attrName, attrValue)
  }
}

function _attributeSupportedByElementOrThrow (attributeName, element) {
  if (!_isAttributeSupportedByElement(attributeName, element)) {
    throw new Error(
      `A "${element.tagName.toLowerCase()}" element do not support the attribute "${attributeName}".`
    )
  }
}

function _isAttributeSupportedByElement (attributeName, element) {
  if (_isDataAttribute(attributeName)) return true
  if (_isAriaAttribute(attributeName)) return true

  return attributeName in element
}

function _isDataAttribute (attributeName) {
  return attributeName.startsWith('data-')
}

function _isAriaAttribute (attributeName) {
  return attributeName.startsWith('aria-')
}

function _addClassNameToElement (className, element) {
  const sanitizedClassNames = _sanitizeClassName(className)

  for (const sanitizedClass of sanitizedClassNames) {
    element.classList.add(sanitizedClass)
  }
}

function _sanitizeClassName (className) {
  return className
    .split(' ')
    .filter(Boolean)
}

function _appendChildrenToElement (children, element) {
  for (const child of children) {
    if (Array.isArray(child)) {
      _appendChildrenToElement(child, element)
    } else {
      _appendChildToElement(child, element)
    }
  }
}

function _appendChildToElement (child, element) {
  const safeChildNode = _createSafeChildNode(child, element)

  element.appendChild(safeChildNode)
}

function _createSafeChildNode (child) {
  if (child instanceof HTMLElement || child instanceof Text) {
    return child
  }

  return document.createTextNode(child)
}
