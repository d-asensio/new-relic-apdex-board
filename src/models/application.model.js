import { uniqueId } from '../utils'

export class Application {
  constructor ({ name, apdex, version, contributors }) {
    this._id = uniqueId('app_')

    this._name = name
    this._apdex = apdex
    this._version = version
    this._contributors = contributors
  }

  get id () { return this._id }
  get name () { return this._name }
  get apdex () { return this._apdex }
  get version () { return this._version }
  get contributors () { return this._contributors }

  compareTo (comparee) {
    return this.apdex - comparee.apdex
  }

  static byApdexInDescOrder (appA, appB) {
    return appB.apdex - appA.apdex
  }
}
