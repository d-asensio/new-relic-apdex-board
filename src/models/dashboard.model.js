export class Dashboard {
  constructor ({ user }) {
    this.user = user

    this._hosts = new Map()
    this._apps = new Map()
  }

  addApp (app) {
    this._apps.set(app.id, app)
  }

  hasApp (appId) {
    return this._apps.has(appId)
  }

  getApp (appId) {
    return this._apps.get(appId)
  }

  addHost (host) {
    this._hosts.set(host.id, host)
  }

  hasHost (hostId) {
    return this._hosts.has(hostId)
  }

  getHost (hostId) {
    return this._hosts.get(hostId)
  }

  getAllHosts () {
    return Array.from(this._hosts.values())
  }
}
