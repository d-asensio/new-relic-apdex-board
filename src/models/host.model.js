export class Host {
  constructor ({ id }) {
    this.id = id
    this._apps = []
  }

  addApp (app) {
    this._insertAppInPosition(app)
  }

  removeApp (app) {
    this._removeAppFromPosition(app)
  }

  getTopApps (nTopApps) {
    return this._apps.slice(0, nTopApps)
  }

  _insertAppInPosition (app) {
    const appInsertionIndex = this._binarySearch(app)
    this._insertAppAtIndex(app, appInsertionIndex)
  }

  _removeAppFromPosition (app) {
    const appIndex = this._binarySearch(app)

    this._indexExistOrThrow(appIndex)
    this._removeAppAtIndex(appIndex)
  }

  _binarySearch (app) {
    let startIndex = 0
    let endIndex = this._apps.length - 1
    let currentIndex

    while (startIndex <= endIndex) {
      currentIndex = (startIndex + endIndex) / 2 | 0
      const iteratingApp = this._apps[currentIndex]

      if (app.compareTo(iteratingApp) < 0) {
        startIndex = currentIndex + 1
      } else if (app.compareTo(iteratingApp) > 0) {
        endIndex = currentIndex - 1
      } else {
        return currentIndex
      }
    }

    return Math.abs(~endIndex)
  }

  _insertAppAtIndex (app, insertionIndex) {
    this._apps.splice(insertionIndex, 0, app)
  }

  _removeAppAtIndex (deletionIndex) {
    this._apps.splice(deletionIndex, 1)
  }

  _indexExistOrThrow (index) {
    if (!this._indexExists(index)) {
      throw new Error(
        'The specified app do not exist in the host.'
      )
    }
  }

  _indexExists (index) {
    return index >= 0 && index < this._apps.length
  }
}
