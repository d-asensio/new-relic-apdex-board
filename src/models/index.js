export * from './application.model.js'
export * from './dashboard.model.js'
export * from './host.model.js'
