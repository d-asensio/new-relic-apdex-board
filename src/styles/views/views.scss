@import "dashboard";
@import "host";
@import "toggle";