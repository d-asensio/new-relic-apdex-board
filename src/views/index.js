export * from './dashboard.view.js'
export * from './host.view.js'
export * from './toggle.view.js'
