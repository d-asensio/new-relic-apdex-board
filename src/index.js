
import './styles/index.scss'

import { DashboardController } from './controllers'
import { DashboardService } from './services'
import { DashboardView } from './views'

const controller = new DashboardController(
  new DashboardService(),
  new DashboardView()
)

function renderApp () {
  const appRoot = document.getElementById('app')
  controller.renderToDom(appRoot)
}

;(async function IIFE () {
  await controller.init()

  if (document.readyState === 'complete') {
    renderApp()
  } else {
    document.addEventListener(
      'DOMContentLoaded',
      () => renderApp()
    )
  }
})()
