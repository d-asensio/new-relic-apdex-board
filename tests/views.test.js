import { Host, Application } from '../src/models'
import { HostView } from '../src/views'

describe('Host view', () => {
  it('is constructed', () => {
    const hostView = new HostView()

    expect(hostView).toBeInstanceOf(HostView)
  })

  it('is created from model', () => {
    const hostView = new HostView()
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const hostElement = hostView.create(host)

    expect(hostElement).toBeInstanceOf(HTMLElement)
  })

  it('is has the proper structure', () => {
    const hostView = new HostView()
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    host.apps.push(
      new Application({
        name: 'A',
        version: 2,
        apdex: 55,
        contributors: [
          'John Doe'
        ]
      })
    )

    const hostElement = hostView.create(host)

    expect(hostElement).toMatchSnapshot()
  })
})
