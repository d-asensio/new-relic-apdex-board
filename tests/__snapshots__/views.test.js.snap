// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard view has the proper structure 1`] = `
<div
  class="Dashboard"
>
  <div
    class="Dashboard__Header"
  >
    <div
      class="Dashboard__HeaderInfo"
    >
      <div
        class="Dashboard__Title"
      >
        Apps by Host
      </div>
      <div
        class="Dashboard__Caption"
      >
        for user 
        some.user@newrelic.com
      </div>
    </div>
    <div
      class="Dashboard__HeaderControls"
    >
      <label
        class="Toggle"
      >
        <input
          type="checkbox"
        />
        <span
          class="Toggle__Mark"
        />
        Show as list
      </label>
    </div>
  </div>
  <div
    class="Dashboard__Content"
  >
    <div
      class="Host"
    >
      <div
        class="Host__Title"
      >
        a2f3d.host.com
      </div>
      <div
        class="Host__AppList"
      >
        <div
          class="Host__App"
        >
          <div
            class="Host__AppApex"
          >
            55
          </div>
          <div
            class="Host__AppName"
          >
            A
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Host view has the proper structure 1`] = `
<div
  class="Host"
>
  <div
    class="Host__Title"
  >
    a2f3d.host.com
  </div>
  <div
    class="Host__AppList"
  >
    <div
      class="Host__App"
    >
      <div
        class="Host__AppApex"
      >
        55
      </div>
      <div
        class="Host__AppName"
      >
        A
      </div>
    </div>
  </div>
</div>
`;

exports[`Toggle view has the proper structure 1`] = `
<label
  class="Toggle"
>
  <input
    type="checkbox"
  />
  <span
    class="Toggle__Mark"
  />
  
</label>
`;
